aws/resource_aws_cloudsearch_service.go

```diff
@@ -0,0 +1,471 @@
+				Type:             schema.TypeString,
+				ValidateFunc:     validateIAMPolicyJson,
+				Required:         true,
+				DiffSuppressFunc: suppressEquivalentAwsPolicyDiffs,
+			},
+		},
+	}
+}
+
+func resourceAwsCloudSearchDomainCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).cloudsearchconn
+
+	input := cloudsearch.CreateDomainInput{
+		DomainName: aws.String(d.Get("domain_name").(string)),
+	}
+
+	output, err := conn.CreateDomain(&input)
+	if err != nil {
+		return fmt.Errorf("%s %q", err, d.Get("domain_name").(string))
+	}
+
+	d.SetId(*output.DomainStatus.ARN)
+	err = resourceAwsCloudSearchDomainUpdate(d, meta)
+
+	return err
+}
+
+func resourceAwsCloudSearchDomainRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).cloudsearchconn
+
+	input := cloudsearch.DescribeIndexFieldsInput{
+		DomainName: aws.String(d.Get("domain_name").(string)),
+	}
+
+	_, err := conn.DescribeIndexFields(&input)
+
+	return err
+}
+
+func resourceAwsCloudSearchDomainUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).cloudsearchconn
+
+	err := updateScalingParameters(d, meta, conn)
+	if err != nil {
+		return err
+	}
+
+	updated, err := defineIndexFields(d, meta, conn)
+	if err != nil {
+		return err
+	}
+
+	err = updateAccessPolicy(d, meta, conn)
+	if err != nil {
+		return err
+	}
+
+	if updated {
+		_, err := conn.IndexDocuments(&cloudsearch.IndexDocumentsInput{
+			DomainName: aws.String(d.Get("domain_name").(string)),
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func updateScalingParameters(d *schema.ResourceData, meta interface{}, conn *cloudsearch.CloudSearch) error {
+	input := cloudsearch.UpdateScalingParametersInput{
+		DomainName: aws.String(d.Get("domain_name").(string)),
+		ScalingParameters: &cloudsearch.ScalingParameters{
+			DesiredInstanceType:     aws.String(d.Get("instance_type").(string)),
+			DesiredPartitionCount:   aws.Int64(int64(d.Get("partition_count").(int))),
+			DesiredReplicationCount: aws.Int64(int64(d.Get("replication_count").(int))),
+		},
+	}
+
+	_, err := conn.UpdateScalingParameters(&input)
+
+	return err
+}
+
+func defineIndexFields(d *schema.ResourceData, meta interface{}, conn *cloudsearch.CloudSearch) (bool, error) {
+	if d.HasChange("indexes") {
+		old := make(map[string]interface{})
+		new := make(map[string]interface{})
+
+		o, n := d.GetChange("indexes")
+
+		for _, ot := range o.([]interface{}) {
+			os := ot.(map[string]interface{})
+			old[os["name"].(string)] = os
+		}
+
+		for _, nt := range n.([]interface{}) {
+			ns := nt.(map[string]interface{})
+			new[ns["name"].(string)] = ns
+		}
+
+		// Handle Removal
+		for k := range old {
+			if _, ok := new[k]; !ok {
+				deleteIndexField(d.Get("domain_name").(string), k, conn)
+			}
+		}
+
+		for _, v := range new {
+			// Handle replaces & additions
+			err := defineIndexField(d.Get("domain_name").(string), v.(map[string]interface{}), conn)
```

Comment:
Can you move the function body here? Since we use it once... 🤷‍♂️ thougths?

Reply:
In my opinion it's ok. Now we have more functions but they all are relatively small and easy to read. If i move the body of defineIndexField into the loop i will have to move deleteIndexField too into the other loop and the function defineIndexFields will grow to a monster.

Have I convinced you? :)